### PR TITLE
Epr refactoring

### PR DESCRIPTION
**Pull request by bvdvecht**
Refactoring of code in `epr_socket.py` and EPR-related code in `builder.py`. These refactors should make it easier to develop new features in the future. All code that depends on these changes, including the examples, have also been updated.

These changes have been made on top of the `builder-refactoring` branch so this PR should be checked only after https://github.com/QuTech-Delft/netqasm/pull/7 has been merged.

### CODE REVIEW

File: netqasm/sdk/build_epr.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    time_unit: TimeUnit = TimeUnit.MICRO_SECONDS
+    max_time: int = 0
+    random_basis_local: Optional[RandomBasis] = None
+    random_basis_remote: Optional[RandomBasis] = None
+    rotations_local: Tuple[int, int, int] = (0, 0, 0)
+    rotations_remote: Tuple[int, int, int] = (0, 0, 0)
+
+
+# Length of NetQASM array for serialized Create Requests.
+SER_CREATE_LEN = 20
```
Comment (QFer) [inline netqasm/sdk/build_epr.py]: To make it more clear that this var must be adjusted make it and put it at the end of the SER_CREATE_xxxx list:
SER_CREATE_LEN = SER_CREATE_IDX_ROTATION_X_REMOTE2  + 1

File: netqasm/sdk/build_epr.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+SER_CREATE_IDX_PROBABLIITY_DIST_REMOTE2 = 13
+SER_CREATE_IDX_ROTATION_X_LOCAL1 = 14
+SER_CREATE_IDX_ROTATION_Y_LOCAL = 15
+SER_CREATE_IDX_ROTATION_X_LOCAL2 = 16
+SER_CREATE_IDX_ROTATION_X_REMOTE1 = 17
+SER_CREATE_IDX_ROTATION_Y_REMOTE = 18
+SER_CREATE_IDX_ROTATION_X_REMOTE2 = 19
+
+# Length of NetQASM array for EPR Keep results.
+SER_RESPONSE_KEEP_LEN = 10
```
Comment (QFer) [inline netqasm/sdk/build_epr.py]: same here if you did the previous

File: netqasm/sdk/build_epr.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+SER_RESPONSE_KEEP_IDX_DIRECTONIALITY_FLAG = 3
+SER_RESPONSE_KEEP_IDX_SEQUENCE_NUMBER = 4
+SER_RESPONSE_KEEP_IDX_PURPOSE_ID = 5
+SER_RESPONSE_KEEP_IDX_REMOTE_NODE_ID = 6
+SER_RESPONSE_KEEP_IDX_GOODNESS = 7
+SER_RESPONSE_KEEP_IDX_GOODNESS_TIME = 8
+SER_RESPONSE_KEEP_IDX_BELL_STATE = 9
+
+# Length of NetQASM array for EPR Measure results.
+SER_RESPONSE_MEASURE_LEN = 10
```
Comment (QFer) [inline netqasm/sdk/build_epr.py]: same here if you did the previous

File: netqasm/sdk/futures.py [hunk 4]
```diff
@@ -334,7 +334,7 @@ def add(
 
         self.builder.subrt_add_pending_commands(commands)
 
-    def _get_load_commands(self, register: operand.Register) -> List[T_Cmd]:
+    def get_load_commands(self, register: operand.Register) -> List[T_Cmd]:
```
Comment (QFer) [inline netqasm/sdk/futures.py:337]: Now public method. I would expect some docstring. When the documentation for netqasm is generated this public method is added and should be documented.

File: netqasm/sdk/futures.py [hunk 5]
```diff
@@ -380,8 +385,17 @@ def _get_access_commands(
         commands.append(access_cmd)
         return commands
 
-    def get_address_entry(self) -> Union[Address, ArraySlice, ArrayEntry]:
-        return parse_address(f"{Symbols.ADDRESS_START}{self._address}[{self._index}]")
+    def get_address_entry(self) -> ArrayEntry:
```
Comment (QFer) [inline netqasm/sdk/futures.py:390]: Public method. I would expect some docstring. Same reason as earlier comment

File: netqasm/sdk/epr_socket.py [hunk 6]
```diff
@@ -710,7 +727,25 @@ def recv_rsp(
...
+                remote_node_id=self.remote_node_id,
+                epr_socket_id=self._epr_socket_id,
+                number=number,
+                post_routine=None,
+                sequential=False,
+            ),
+        )
+        return qubits
+
+    def recv_rsp_with_info(
```
Comment (QFer) [inline netqasm/sdk/epr_socket.py:757]: some public methods have nice docstring. Why not this one?

File: netqasm/sdk/epr_socket.py [hunk 7]
```diff
@@ -661,11 +660,29 @@ def recv_keep(
                 sequential=sequential,
             ),
         )
+        return qubits
+
+    def recv_keep_with_info(
```
Comment (QFer) [inline netqasm/sdk/epr_socket.py:673]: some public methods have nice docstring. Why not this one?

File: netqasm/sdk/epr_socket.py [hunk 8]
```diff
@@ -228,8 +229,30 @@ def post_create(conn, q, pair):
                 max_time=max_time,
             ),
         )
+        return qubits
 
-    def get_rotations_from_basis(self, basis: EPRMeasBasis) -> Tuple[int, int, int]:
+    def create_keep_with_info(
```
Comment (QFer) [inline netqasm/sdk/epr_socket.py:234]: Public methods should have docstring.

File: netqasm/sdk/builder.py [hunk 9]
```diff
@@ -1787,15 +1682,25 @@ def sdk_try_context(
...
-    # ) -> None:
-    #     try:
-    #         pre_commands = self.subrt_pop_all_pending_commands()
-    #         yield
-    #     finally:
-    #         body_commands = self.subrt_pop_all_pending_commands()
-    #         commands = pre_commands + body_commands
-    #         self.subrt_add_pending_commands(commands)
+    @contextmanager
+    def sdk_create_epr_context(
```
Comment (QFer) [inline netqasm/sdk/builder.py:1705]: Public methods should have docstring.

File: netqasm/sdk/builder.py [hunk 10]
```diff
@@ -1680,20 +1570,25 @@ def sdk_epr_rsp_recv(
             qubit_ids_array, ent_results_array, wait_all, params
         )
 
-        return qubit_futures
+        epr_results = deserialize_epr_keep_results(params.number, ent_results_array)
+        return qubit_futures, epr_results
 
-    def sdk_create_epr_keep(self, params: EntRequestParams) -> List[Qubit]:
+    def sdk_create_epr_keep(
```
Comment (QFer) [inline netqasm/sdk/builder.py:1582]: Public methods should have docstring. We generate read the doc for this don't we? Some other methods also in this module have no docstring

**Review by QFer — CHANGES_REQUESTED** [review]

File: netqasm/sdk/build_epr.py (hunk 1 above)
Comment (bvdvecht) [reply netqasm/sdk/build_epr.py]: done

File: netqasm/sdk/build_epr.py (hunk 3 above)
Comment (bvdvecht) [reply netqasm/sdk/build_epr.py]: done

File: netqasm/sdk/futures.py (hunk 4 above)
Comment (bvdvecht) [reply netqasm/sdk/futures.py:337]: added docstring

File: netqasm/sdk/epr_socket.py (hunk 6 above)
Comment (bvdvecht) [reply netqasm/sdk/epr_socket.py:757]: added

File: netqasm/sdk/builder.py (hunk 10 above)
Comment (bvdvecht) [reply netqasm/sdk/builder.py:1582]: added docstrings (also for other methods)

**Review by QFer — APPROVED** [review]

**Pull request merged**